content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,661 @@
+            14 ┊     0.00% ┊ export "__heap_base"
+            13 ┊     0.00% ┊ export "__data_end"
+            12 ┊     0.00% ┊ custom section 'producers' headers
+             9 ┊     0.00% ┊ export "memory"
+             9 ┊     0.00% ┊ add
+...
+```
+
+It’s now clearly visible that all main contributors to the module size are custom sections, which — by definition — are not relevant for the execution of the module. Their names imply that they contain information that is used for debugging, so the fact that these sections are emitted for a `--release` build is somewhat surprising. It seems related to a long-standing bug, where _our_ code is compiled without debug symbols, but the pre-compiled standard library on our machine still has debug symbols.
+
+ To address this we add another line to our `Cargo.toml`:
+
+```toml
+[profile.release]
+strip = true
+```
+
+There is also tool provided by [wabt] called `wasm-strip` which removes all custom sections. The benefit of `wasm-strip` is that it allows you to keep specific custom sections around, like the `name` section that gives all your Wasm functiosn readable names (that is, once my [PR][wasm-strip pr] lands):
+
+```
+$ wasm-strip -k target/wasm32-unknown-unknown/release/my_project.wasm
+```
+
+After stripping, we are left with a module of a whopping 116B. Disassembling it shows that the only function in that module is called `add` and executes `(f64.add (local.get 0) (local.get 1))`, which means the Rust compiler was able to emit optimal code. Of course, staying on top of binary size gets more complicated with a growing code base.
+
+
+### Sneaky bloat
+
+I have seen a couple of complaints online about how big WebAssembly modules create by Rust are that do a seemingly small job. In my experience, there are three reasons why Rust emits surprisingly big WebAssembly binaries:
+
+* Debug build (i.e. forgetting to pass `--release` to `Cargo`)
+* Debug symbols (i.e. forgetting to run `wasm-strip`)
+* String formatting and panics 
+
+We have looked at the first two. Let’s take a closer look at the last one. This innocuous line of code compiles to 18KB of WebAssembly:
+
+```rust
+static PRIMES: &[i32] = &[2, 3, 5, 7, 11, 13, 17, 19, 23];
+
+#[no_mangle]
+fn nth_prime(n: usize) -> i32 {
+    PRIMES[n]
+}
+```
+
+Okay, maybe not so innocuous after all. You might already know what's going on here. 
+
+### Panicking
+
+A quick look at `twiggy` shows that the main contributors to the Wasm module size are functions related to string formatting and panicking (before LTO, there was also code for memory management in there). And that makes sense! The parameter `n` is unsanitized and used to index an array. Rust has no choice but to inject bounds checks. If a bounds check fails, Rust panics.
+
+One way to handle this is to do the bounds checking ourselves. Rust's compiler is really good at proving whether the code can end up accessing out of bounds.
```

Comment:
Btw, looks like, unlike surrounding paragraphs, this one doesn't mention the size after changes / how much removing the panics saves us.